glean-core/src/metrics/dual_labeled_counter.rs

```diff
@@ -0,0 +1,375 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::borrow::Cow;
+use std::char;
+use std::collections::HashMap;
+use std::io::Read;
+use std::sync::{Arc, Mutex};
+
+use crate::common_metric_data::{CommonMetricData, CommonMetricDataInternal, DynamicLabelType};
+use crate::error_recording::{record_error, test_get_num_recorded_errors, ErrorType};
+use crate::metrics::{CounterMetric, Metric, MetricType};
+use crate::Glean;
+
+const MAX_LABELS: usize = 16;
+const OTHER_LABEL: &str = "__other__";
+const MAX_LABEL_LENGTH: usize = 111;
+const RECORD_SEPARATOR: char = '\x1E';
+
+/// A dual labled metric, mirrors the Legacy keyed-categorical type.
+///
+/// Dual labled metrics allow recording multiple sub-metrics of the same type, in relation
+/// to two dimensions rather than the single label provided by the standard labeled type.
+#[derive(Debug)]
+pub struct DualLabeledCounterMetric {
+    keys: Option<Vec<Cow<'static, str>>>,
+    categories: Option<Vec<Cow<'static, str>>>,
+    /// Type of the underlying metric
+    /// We hold on to an instance of it, which is cloned to create new modified instances.
+    counter: CounterMetric,
+
+    /// A map from a unique ID for the dual labeled submetric to a handle of an instantiated
+    /// metric type.
+    dual_label_map: Mutex<HashMap<(String, String), Arc<CounterMetric>>>,
+}
+
+// TODO figure out how to measure the sizeof these metrics
+// impl ::malloc_size_of::MallocSizeOf for DualLabeledCounterMetric {
+//     fn size_of(&self, ops: &mut malloc_size_of::MallocSizeOfOps) -> usize {
```

Comment:
Can you implement this as just returning `0`? Then at least for now it should work in m-c.
I can fix that up in a followup.